### PR TITLE
Improve llvm not found error

### PR DESCRIPTION
**Pull request by esc**
This PR continues on from the starter patch at:

https://github.com/stuartarchibald/llvmlite/commit/e7dd3195051709c1b1a7e47edf760de5ecbef7dc

The goal is to address: https://numba.discourse.group/t/fixing-filenotfounderror-errno-2-no-such-file-or-directory-llvm-config/510

### CODE REVIEW

**Review by sklam — APPROVED** [review]
The update doc now clearly explains the problems and what to do when there are installation problem. I have also manually tested error message from build.py. Hopefully, this will help user resolve any installation problems better.

File: docs/source/admin-guide/install.rst [hunk 1]
```diff
@@ -2,31 +2,139 @@
...
+The Numba/llvmlite stack consists of the following major components:
+
+* *Numba* is the compiler package, this depends on llvmlite.
+* *llvmlite* is a lightweight binding package to the LLVM APIs, it depends on LLVM.
+* *LLVM*  is the JIT compiler framework for producing executable code from various
+  inputs.
+
+All components must be compiled in order to be used. And, since each component
+on the stack depends on the previous one, you need to compile LLVM in order to
+compile llvmlite in order to compile Numba. The LLVM package has a significant
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:14]:
```suggestion
compile llvmlite in order to compile Numba. The LLVM package is a significant
```

File: docs/source/admin-guide/install.rst [hunk 2]
```diff
@@ -2,31 +2,139 @@
...
 
 Pre-built binaries
 ==================
 
-Building LLVM for llvmlite is challenging, so we *strongly* recommend
-installing a binary package where we have built and tested everything for you.
-Official conda packages are available in the Anaconda_ distribution::
+As mentioned above, Building LLVM for llvmlite is challenging, so we *strongly*
+recommend installing a binary package where we have built and tested everything
+for you.
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:23]:
```suggestion
As mentioned above, building LLVM for llvmlite is challenging. Installing a binary
package that has been built and tested is *strongly* recommend.
```

File: docs/source/admin-guide/install.rst [hunk 3]
```diff
@@ -2,31 +2,139 @@
...
+the Numba_ development channel on `Anaconda Cloud <https://anaconda.org/numba>`_::
 
-    conda install --channel numba llvmlite
+    conda install -c numba/label/dev llvmlite
 
 Binary wheels are also available for installation from PyPI_::
 
     pip install llvmlite
 
+Development releases for binary wheels are not made available.
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:38]:
```suggestion
Development releases of binary wheels are not made available.
```

File: docs/source/admin-guide/install.rst [hunk 4]
```diff
@@ -2,31 +2,139 @@
...
+
+Contrary to what you might expect, the llvmlite packages built by the Numba
+maintainers do *not* use any LLVM shared libraries that may be present on the
+system, and/or in the Conda environment. The parts of LLVM required by llvmlite
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:42]:
```suggestion
Contrary to what might be expected, the llvmlite packages built by the Numba
maintainers do *not* use any LLVM shared libraries that may be present on the
system, and/or in the Conda environment. The parts of LLVM required by llvmlite
```

File: docs/source/admin-guide/install.rst [hunk 5]
```diff
@@ -2,31 +2,139 @@
...
+are statically linked at build time.  As a result, installing llvmlite from a
+binary package does not also require the end user to install LLVM.  (For more
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:44]:
```suggestion
are statically linked at build time.  As a result, installing llvmlite from a
binary package from the Numba channel does not also require the end user to
install LLVM.  (For more
```

File: docs/source/admin-guide/install.rst [hunk 6]
```diff
@@ -2,31 +2,139 @@
...
+details on the reasoning behind this, see: :ref:`faq_why_static`). Note however
+also that llvmlite packages compiled by other parties, e.g. conda-forge may
+split this into and ``llvmlite`` and ``llvm`` package.
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:47]:
```suggestion
also that llvmlite packages compiled by other parties, e.g. conda-forge may
split this into and ``llvmlite`` and ``llvm`` package and link dynamically.
```

File: docs/source/admin-guide/install.rst [hunk 7]
```diff
@@ -2,31 +2,139 @@
...
+-----------
+
+The Numba maintainers ship binary wheels:
+
+  * Numba wheels (``x86*`` architectures)
+  * llvmlite wheels (``x86*`` architectures)
+
+Note that the llvmlite wheels are statically linked against LLVM, as per the
+conda packages on the Numba channel. This mitigates the need for a LLVM based
+binary wheel. Not also that this, as of version 0.36 does not include the
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:87]:
```suggestion
binary wheel. Note also that this, as of version 0.36, does not include the
```

File: docs/source/admin-guide/install.rst [hunk 8]
```diff
@@ -2,31 +2,139 @@
...
+``aarch64`` architectures, for example installation on a RaspberryPis is not
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:88]:
```suggestion
``aarch64`` architectures, for example installation on a Raspberry Pi is not
```

File: docs/source/admin-guide/install.rst [hunk 9]
```diff
@@ -2,31 +2,139 @@
...
+The Numba maintainers ship an ``sdist`` for:
+
+  * Numba
+  * llvmlite
+
+Note that there is no ``sdist`` provided for LLVM. If you try and build ``llvmlite``
+from ``sdist`` you will need to bootstrap the package with your own appropriate
+LLVM.
+
+How this ends up being a problem.
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:100]: Should this be a sub-title?

File: docs/source/admin-guide/install.rst [hunk 10]
```diff
@@ -2,31 +2,139 @@
...
+message, which included here verbatim for future reference::
+
+    FileNotFoundError: [Errno 2] No such file or directory: 'llvm-config'
+
+Things to "fix" it...
+
+1. If you are using ``pip < 19.0`` and on ``x86*``, then update it if you can, this will
+   let you use the ``manylinux2010`` binary wheels.
+
+2. If you are on an unsupported architecture, for example RaspberryPis, please
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:121]:
```suggestion
2. If you are on an unsupported architecture, for example Raspberry Pi, please
```

File: docs/source/admin-guide/install.rst [hunk 11]
```diff
@@ -2,31 +2,139 @@
...
+   use ``conda`` if you have that available.
+
+3. Otherwise: you will probably need to build from source, this means providing
+   an LLVM. If you have conda available you could use this to bootstrap the
+   installation with a working ``llvm``/`llvmdev` package. If you have a system
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:126]:
```suggestion
   installation with a working ``llvm``/``llvmdev`` package. If you have a system
```

File: docs/source/admin-guide/install.rst [hunk 12]
```diff
@@ -2,31 +2,139 @@
...
+   LLVM that's of an appropriate version you could use that. See
+   https://llvmlite.readthedocs.io/en/latest/admin-guide/install.html#building-manually
+   and in particular note the use of the ``LLVM_CONFIG`` environment variable
+   for specifying where your LLVM install is.
+
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:131]: This section and the next `What not to do...` section seem to conflict? This section suggests that if you have a system LLVM that's appropriate you could use it, but the next suggests to not use a package managed one (i.e. system). Perhaps the next second should be called "What to be aware of if using a system provided LLVM" and note that a) might not work b) unlikely patches in a way all the tests will pass c) may be harder to get help.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for putting this into the docs, it's asked so often it's great to have a reference for it. Few minor comments to resolve else looks good.

File: docs/source/admin-guide/install.rst [hunk 13]
```diff
@@ -2,31 +2,139 @@
...
+At compile time the ``llvmdev`` and ``llvm`` packages are used to build llvmlite and
+llvmlite's dynamic libraries are dynamically linked against the libraries in the
+``llvm`` meta-package. This means at runtime ``llvmlite`` depends on the ``llvm``
+package which has the LLVM shared libraries in it (it's actually a package
+called ``libllvm`` that contains the DSOs, but the ``llvm`` package is referred to
+so as to get the ``run_exports``).
+
+Using `pip`
+-----------
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:78]:
```suggestion
Using ``pip``
-------------
```

File: docs/source/admin-guide/install.rst (hunk 9 above)
Comment (esc) [reply docs/source/admin-guide/install.rst:100]: probably, but then the later `Things to "fix" it...` should be one also, perhaps?

File: docs/source/admin-guide/install.rst (hunk 12 above)
Comment (esc) [reply docs/source/admin-guide/install.rst]: OK, I have reworded the section, please take another look.

File: docs/source/admin-guide/install.rst [hunk 14]
```diff
@@ -11,16 +11,15 @@ The Numba/llvmlite stack consists of the following major components:
...
 size and may take significant time (magnitude, roughly an hour) and skill to
 compile depending on the platform.
 
 Pre-built binaries
 ==================
 
-As mentioned above, Building LLVM for llvmlite is challenging, so we *strongly*
-recommend installing a binary package where we have built and tested everything
-for you.
+As mentioned above, building LLVM for llvmlite is challenging. Installing a binary
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:21]: line wrap?

File: docs/source/admin-guide/install.rst [hunk 15]
```diff
@@ -114,27 +115,32 @@ message, which included here verbatim for future reference::
...
-   LLVM that's of an appropriate version you could use that. See
-   https://llvmlite.readthedocs.io/en/latest/admin-guide/install.html#building-manually
+   installation with a working ``llvm``/``llvmdev`` package. Learn more about
+   compiling from source in the section on `Building manually`_ below.
    and in particular note the use of the ``LLVM_CONFIG`` environment variable
    for specifying where your LLVM install is.
 
-What not to do...
+What to be aware of when using a pre-built LLVM package
+.......................................................
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:134]:
```suggestion
What to be aware of when using a system provided LLVM package.
..............................................................
```
I think `pre-built` would include one from e.g. `conda`?

File: docs/source/admin-guide/install.rst [hunk 16]
```diff
@@ -114,27 +115,32 @@ message, which included here verbatim for future reference::
...
+
+When using a system provided LLVM package, there are a number of things that
+could go wrong:
 
-1. Do not install a  pre-built LLVM package through your package manager (apt,
-   yum, brew...) or otherwise. The LLVM contained within almost certainly
-   doesn't carry the patches needed for running the Numba/llvmlite stack. Doing
-   this anyway may result in erroneous behaviour at runtime.
+1. The LLVM package may not work with Numba/llvmlite at all
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:139]:
```suggestion
1. The LLVM package may not work with Numba/llvmlite at all.
```

File: docs/source/admin-guide/install.rst [hunk 17]
```diff
@@ -114,27 +115,32 @@ message, which included here verbatim for future reference::
...
+2. If it does it is unlikely the carry the correct patches for Numba/llvmlite
+   to work appropriately.
```
Comment (stuartarchibald) [inline docs/source/admin-guide/install.rst:141]:
```suggestion
2. If it does work to some degree it is unlikely the carry the correct patches
   for Numba/llvmlite to work entirely correctly.
```

File: docs/source/admin-guide/install.rst (hunk 12 above)
Comment (stuartarchibald) [reply docs/source/admin-guide/install.rst]: Great, thanks.

File: docs/source/admin-guide/install.rst (hunk 9 above)
Comment (esc) [reply docs/source/admin-guide/install.rst:100]: done

File: docs/source/admin-guide/install.rst (hunk 14 above)
Comment (esc) [reply docs/source/admin-guide/install.rst]: sure

**Comment by esc** [comment]
OK; think it's all been addressed again.

**Comment by stuartarchibald** [comment]
> OK; think it's all been addressed again.

Brilliant, thanks for the fixes.

**Review by stuartarchibald — APPROVED** [review]
Thanks for all the fixes, looks good!

**Pull request merged**